src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,161 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> tempNode = new Node<>(null, value, null);
+        if (size == 0) {
+            head = tempNode;
+        } else {
+            tail.next = tempNode;
+            tempNode.prev = tail;
+        }
+        tail = tempNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == 0) {
+            if (size == 0) {
+                Node<T> current = new Node<>(null, value, null);
+                head = current;
+                tail = current;
+            } else {
+                Node<T> newHead = new Node<>(null, value, head);
+                newHead.next = head;
+                head.prev = newHead;
+                head = newHead;
+            }
+        } else if (index == size) {
+            Node<T> newTail = new Node<>(tail, value, null);
+            newTail.prev = tail;
+            tail.next = newTail;
+            tail = newTail;
+        } else {
+            Node<T> nodeByIndex = getNode(index);
+            Node<T> middleCurrent = new Node<>(nodeByIndex.prev, value, nodeByIndex);
+            nodeByIndex.prev.next = middleCurrent;
+            middleCurrent.prev = nodeByIndex.prev;
+            nodeByIndex.prev = middleCurrent;
+            middleCurrent.next = nodeByIndex;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t:list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index == 0) {
+            final T oldValue = head.item;
+            Node<T> newHead = new Node<>(null, value, head.next);
+            newHead.next = head.next;
+            head.next.prev = newHead;
+            head = newHead;
+            return oldValue;
+        }
+        Node<T> nodeToChange = getNode(index);
+        Node<T> node = new Node<>(nodeToChange.prev, value, nodeToChange.next);
+        nodeToChange.prev.next = node;
+        nodeToChange.next.prev = node;
+        return nodeToChange.item;
```

Comment:
Don't create a new node.
1. get the node (line 74)
2. save old value in a local var (will be returned)
3. reassign item of the node to the new value
4. return old value
We don't need to handle index == 0 separately